svcdef/svcdef.go

```diff
@@ -0,0 +1,347 @@
+// type or an Enum type. In the resulting Go code, a basetype will be
+// represented as an ast.Ident, while a key that is a Message or Enum type will
+// be represented as an *ast.StarExpr which references an ast.Ident.
+func NewMap(m ast.Expr) (*Map, error) {
+	rv := &Map{
+		KeyType:   &FieldType{},
+		ValueType: &FieldType{},
+	}
+	mp := m.(*ast.MapType)
+	// KeyType will always be an ast.Ident, ValueType may be an ast.Ident or an
+	// ast.StarExpr->ast.Ident
+	key := mp.Key.(*ast.Ident)
+	rv.KeyType.Name = key.Name
+	var keyFollower func(ast.Expr)
+	keyFollower = func(e ast.Expr) {
+		switch ex := e.(type) {
+		case *ast.Ident:
+			rv.ValueType.Name = ex.Name
+		case *ast.StarExpr:
+			rv.ValueType.StarExpr = true
+			keyFollower(ex.X)
+		}
+	}
+	keyFollower(mp.Value)
+
+	return rv, nil
+}
+
+// NewService returns a new Service struct derived from an *ast.TypeSpec with a
+// Type of *ast.InterfaceType.
+func NewService(s *ast.TypeSpec) (*Service, error) {
+	rv := &Service{
+		Name: s.Name.Name,
+	}
+	asvc := s.Type.(*ast.InterfaceType)
+	for _, m := range asvc.Methods.List {
+		nmeth, err := NewServiceMethod(m)
+		if err != nil {
+			return nil, errors.Wrapf(err, "Couldn't create service method %q of service %q", m.Names[0].Name, rv.Name)
+		}
+		rv.Methods = append(rv.Methods, nmeth)
+	}
+	return rv, nil
+}
+
+// NewServiceMethod returns a new ServiceMethod derived from the provided
+// *ast.Field. The given *ast.Field is intended to have a Type of *ast.FuncType
+// from an *ast.InterfaceType's Methods.List attribute. Providing an *ast.Field
+// with a different structure may return an error.
+func NewServiceMethod(m *ast.Field) (*ServiceMethod, error) {
+	rv := &ServiceMethod{
+		Name: m.Names[0].Name,
+	}
+	ft, ok := m.Type.(*ast.FuncType)
+	if !ok {
+		return nil, errors.New("Provided *ast.Field.Type is not of type *ast.FuncType; cannot proceed")
+	}
+
+	input := ft.Params.List
+	output := ft.Results.List
+
+	// Zero'th param of a serverMethod is Context.context, while first param is
+	// this methods RequestType
+	rq := input[1]
+	rs := output[0]
+
+	makeFieldType := func(in *ast.Field) (*FieldType, error) {
+		star, ok := in.Type.(*ast.StarExpr)
+		if !ok {
+			return nil, errors.New("could not create FieldType, in.Type is not *ast.StarExpr")
+		}
+		ident, ok := star.X.(*ast.Ident)
+		if !ok {
+			return nil, errors.New("could not create FieldType, star.Type is not *ast.Ident")
+		}
+		return &FieldType{
+			Name:     ident.Name,
+			StarExpr: true,
+		}, nil
+	}
+
+	var err error
+	rv.RequestType, err = makeFieldType(rq)
+	if err != nil {
+		return nil, errors.Wrapf(err, "RequestType creation of service method %q failed", rv.Name)
+	}
+	rv.ResponseType, err = makeFieldType(rs)
+	if err != nil {
+		return nil, errors.Wrapf(err, "ResponseType creation of service method %q failed", rv.Name)
+	}
+
+	return rv, nil
+}
+
+// NewField returns a Field struct with information distilled from an
```

Comment:
Another comment that is not helpful in the docs, possibly bring the chart up from below. Or remove or give context of why an error might be returned.

![2016-10-13-142114_631x138_scrot](https://cloud.githubusercontent.com/assets/19537819/19367875/0f273bfe-9152-11e6-937f-439bf5069727.png)